EIPS/eip-6224.md

```diff
@@ -0,0 +1,244 @@
+- It SHOULD NOT be possible to add zero addresses to the `ContractsRegistry`.
+- The `ContractsRegistry` MUST use the `IDependant` interface in the `injectDependencies` method.
+
+### Dependant
+
+The `Dependant` contract is the one that depends on other contracts present in the system. In order to support dependency injection mechanism, the dependant contract MUST implement the following interface:
+
+```solidity
+pragma solidity ^0.8.0;
+
+interface IDependant {
+   /**
+    *  @notice The function that is called from the ContractsRegistry (or factory) to inject dependencies.
+    *  @param contractsRegistry the registry to pull dependencies from
+    *
+    *  The Dependant MUST perform a dependency injector access check to this method
+    */
+   function setDependencies(address contractsRegistry) external;
+
+   /**
+    *  @notice The function that sets the new dependency injector.
+    *  @param injector the new dependency injector
+    *
+    *  The Dependant MUST perform a dependency injector access check to this method
+    */
+   function setInjector(address injector) external;
+
+   /**
+    *  @notice The function that gets the current dependency injector
+    *  @return the current dependency injector
+    */
+   function getInjector() external view returns (address);
+}
+```
+
+- The `Dependant` contract MUST pull its dependencies in the `setDependencies` method from the passed `contractsRegistry` address.
+- The `Dependant` contract MAY store the dependency injector address in the special slot `0x3d1f25f1ac447e55e7fec744471c4dab1c6a2b6ffb897825f9ea3d2e8c9be583` (obtained as `bytes32(uint256(keccak256("eip6224.dependant.slot")) - 1)`).
+
+
+## Rationale
+
+The EIP proposes a way to manage the system's smart contracts and maintain a clear usage of their dependencies. There are a few design decisions that have to be specified explicitly:
+
+### ContractsRegistry Rationale
+
+**Usage**
+
+The `ContractsRegistry` SHOULD be the topmost smart contract in the system. It SHOULD NOT depend on other contracts. The extensions of this EIP SHOULD add proper access control check to the described non-view methods. 
+
+The `getContract` and `getImplementation` methods SHOULD revert if the nonexisting contracts are asked. This logic MAY be used to ensure that proper dependencies are provided to the dependants.
+
+The `ContractsRegistry` MAY be set behind the proxy to enable runtime addition of custom methods. Applications MAY also leverage the pattern to develop custom tree-like `ContractsRegistry` data structures.
+
+**Contracts identifier**
+
+The `string` contracts identifier is chosen over the `uint256` and `bytes32` to maintain code readability and reduce the human-error chances when interacting with the `ContractsRegistry`. Being the topmost smart contract, it MAY be typical for the users to interact with it via block explorers or DAOs. Clarity, not gas usage, is a factor of choice.
```

Comment:
```suggestion
The `string` contracts identifier is chosen over the `uint256` and `bytes32` to maintain code readability and reduce the human-error chances when interacting with the `ContractsRegistry`. Being the topmost smart contract, it MAY be typical for the users to interact with it via block explorers or DAOs. Clarity was prioritized over gas usage.
```